### PR TITLE
Fix RESTMapper to dynamically rediscover APIs

### PR DESCRIPTION
**Pull request by adracus**
**What this PR does / why we need it**:
The gardener-controller-manager had the issue that if the gardener-apiserver came up too late, the discovery information of the controller runtime was outdated, leading to resource not found issues. By using the cached discovery client, we can fix this issue since it allows dynamic rediscovering of resources.

Co-Authored-By: Tim Usner <tim.usner@sap.com>

**Special notes for your reviewer**:

**Release note**:

```noteworthy operator
The `gardener-controller-manager` now uses a cached discovery client for the `Garden` cluster that can be configured via entries in the `ControllerManagerConfiguration`. Also, the previous two environment variables `KUBECONFIG` and `GARDENER_KUBECONFIG` have been merged into a single `KUBECONFIG` variable. This means Gardener will communicate to only a single `Garden` cluster, being more in line with other Kubernetes controllers.
```

### CODE REVIEW

**Review by timuthy — DISMISSED** [review]
/lgtm. May you revert `5db09310296ce9c6c348fd241c0fa0cf3129dee5` as part of this PR?

File: example/20-componentconfig-gardener-controller-manager.yaml [hunk 1]
```diff
@@ -44,6 +44,11 @@ leaderElection:
   renewDeadline: 10s
   retryPeriod: 2s
   resourceLock: configmaps
+# `discovery` allows specifying options for the cached discovery client.
+# discovery:
```
Comment (rfranzke) [inline example/20-componentconfig-gardener-controller-manager.yaml:48]: Can you please implement the possibility to configure that in the Gardener Helm chart?

**Review by rfranzke — COMMENTED** [review]
Also, please revert `5db09310296ce9c6c348fd241c0fa0cf3129dee5` as suggest by @timuthy

**Comment by rfranzke** [comment]
@mvladev can you please take a look?

**Comment by adracus** [comment]
Reverted `5db09310296ce9c6c348fd241c0fa0cf3129dee5` and made the switches configurable via helm settings.

File: pkg/controllermanager/apis/config/types.go [hunk 2]
```diff
@@ -30,14 +30,12 @@ type ControllerManagerConfiguration struct {
 	// ClientConnection specifies the kubeconfig file and client connection
 	// settings for the proxy server to use when communicating with the gardener-apiserver.
 	ClientConnection apimachineryconfig.ClientConnectionConfiguration
-	// GardenerClientConnection specifies the kubeconfig file and client connection
-	// settings for the garden-apiserver.
-	// +optional
-	GardenerClientConnection *apimachineryconfig.ClientConnectionConfiguration
```
Comment (mvladev) [inline pkg/controllermanager/apis/config/types.go:36]: Why you have removed that? The use case is that you might talk directly to the `gardener-apiserver` without going through the `kube-aggregator`

Comment (adracus) [reply pkg/controllermanager/apis/config/types.go:36]: Yes we know, however this has never ever been used in practice - also, in general, the API server should rather appear as one entity other than as multiple, separated ones. I think after this long time of maintaining this without any helpful outcome, it is now the time to remove this.

Comment (rfranzke) [reply pkg/controllermanager/apis/config/types.go:36]: @mvladev any follow up? Or OK to remove it?

Comment (adracus) [reply pkg/controllermanager/apis/config/types.go:36]: @mvladev any comments?

Comment (mvladev) [reply pkg/controllermanager/apis/config/types.go:36]: I'm file with removal so /lgtm

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**